### ISSUE TITLE
Questions about distributed training on multiple machines

### ISSUE THREAD

**Issue by LakeCarrot** [apache/incubator-mxnet#7011]
Hi,

I tried to run the [mnist](https://github.com/dmlc/mxnet/blob/master/example/image-classification/train_mnist.py) example provided in mxnet on a set of ssh-able machines. I run the program using the following command.

`python ../../tools/launch.py -n 2 -H hosts \ `
`python ./train_mnist.py --network mlp --num-epochs 1 --kv-store dist_sync`

Everything seems work well. But I have two concerns. 
1) How can I get a global view this distributed process? The log generated by the workers is similar to what I got if I run the same workload under the single machine setting. Even if I set the `PS_VERBOSE=1`, I can only know that all the servers and workers have registered with the scheduler and can talk with each other. Take the mnist as an example, there are 60000 samples in the training set, how can I know these workers are coorperating together to get the work done instead of running seperately?

FYI, here is the log generated by one of the workers when run the distributed training example. 
```
[19:54:00] src/van.cc:75: Bind to role=worker, ip=10.0.2.216, port=38145, is_recovery=0
[19:54:24] src/van.cc:281: W[9] is connected to others
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='dist_sync', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=1, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): yann.lecun.com
DEBUG:urllib3.connectionpool:http://yann.lecun.com:80 "GET /exdb/mnist/train-labels-idx1-ubyte.gz HTTP/1.1" 200 28881
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): yann.lecun.com
DEBUG:urllib3.connectionpool:http://yann.lecun.com:80 "GET /exdb/mnist/train-images-idx3-ubyte.gz HTTP/1.1" 200 9912422
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): yann.lecun.com
DEBUG:urllib3.connectionpool:http://yann.lecun.com:80 "GET /exdb/mnist/t10k-labels-idx1-ubyte.gz HTTP/1.1" 200 4542
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): yann.lecun.com
DEBUG:urllib3.connectionpool:http://yann.lecun.com:80 "GET /exdb/mnist/t10k-images-idx3-ubyte.gz HTTP/1.1" 200 1648877
INFO:root:Epoch[0] Batch [100]	Speed: 259.32 samples/sec	accuracy=0.808014
INFO:root:Epoch[0] Batch [200]	Speed: 252.99 samples/sec	accuracy=0.928125
INFO:root:Epoch[0] Batch [300]	Speed: 255.75 samples/sec	accuracy=0.942031
INFO:root:Epoch[0] Batch [400]	Speed: 259.36 samples/sec	accuracy=0.947187
INFO:root:Epoch[0] Batch [500]	Speed: 258.24 samples/sec	accuracy=0.953438
INFO:root:Epoch[0] Batch [600]	Speed: 255.40 samples/sec	accuracy=0.963281
INFO:root:Epoch[0] Batch [700]	Speed: 254.65 samples/sec	accuracy=0.967031
INFO:root:Epoch[0] Batch [800]	Speed: 256.54 samples/sec	accuracy=0.960625
INFO:root:Epoch[0] Batch [900]	Speed: 258.10 samples/sec	accuracy=0.968594
INFO:root:Epoch[0] Train-accuracy=0.972551
INFO:root:Epoch[0] Time cost=233.724
INFO:root:Epoch[0] Validation-accuracy=0.966063
[19:58:44] src/./zmq_van.h:46: W[9] is stopping
[19:58:44] src/van.cc:170: W[9] is stopped
```

2) I found that it's quite convienent to run the workload on a set of cpu machines or a set of gpu machines. However, I wonder is there a way to run the workload on a mixed of cpu and gpu workers? For example, there are 5 workers, 2 of them have 2 CPUs on each of them, others have 3 GPUs on each of them. Is that possbile for us to launch the distributed training task on these 5 machines? Could you please give a pointer of how to do that?

Thanks in advance!
Bo

**Comment by madjam** [comment]
1. When running in dist_sync mode training would come to a halt if one of the workers or servers is down. So that is a hint. But your observation may still be valid. What information do you think would be useful for diagnostic purposes.


2. This probably does not answer your question, but I'm curious to know why you want to do this? Are you running in a shared environment? You probably don't want to run in sync mode here otherwise your gpu workers will sit idle most of the time waiting for cpu workers. Even in async mode, there is a marked asymmetry between the rates at which gpus and cpus can process input data.

**Comment by LakeCarrot** [comment]
@madjam Thanks for your comments!
1. If we consider data parallelism, as each worker works on one partition of the data. If we can monitor the overall data processed by all the workers (represented by the total number of processed batches for example) and also get the progress each worker has made (represented by the local processed batches since the work launched) in the runtime, then it will be more straightforward to show the global view the distributed learning.
2. To be frank, I am new to deep learning, so the following concerns are mostly hypothetical. 
One of the scenario I can come up with is what if there are not enough computation resource for you to do the computation. Although intuitively, the more GPUs we have the quicker the learning process, but we always have a limited number of GPU. Under this circumstances, if we can leverage the computation capacity of the idle CPU cluster in our system, we will get extra performance gain. 
Another concern would be is that possible some of the computation is processed faster on CPU? I know that computation like matrix multiplication will be extremely fast if we use GPU. But I am not sure that is always the case for all kinds of computation in deep learning algorithms.
Do these two concerns make sense to you?

**Comment by madjam** [comment]
1. [Module.fit](http://mxnet.io/api/python/module.html#mxnet.module.BaseModule.fit) lets your supply callbacks that will be invoked when events of interest such as batch end, epoch end occur. You can supply a logging call back to have it print out stats.
2. Operations that consume lots of FLOPS are better run on GPUs, assuming you have them available. In some cases the discrepancy in training times (between CPU and GPU) is so big that you probably are better off waiting for a GPU to become available. You could conceivably use a cluster of CPUs if you are not training the entire network i.e. taking a pre-trained model and fine tuning it with new labeled data. I do think your questions/concerns are valid. It is just that CPUs and GPUs are optimized for different types of workloads.

**Comment by LakeCarrot** [comment]
@madjam Thanks for comments! That helps a lot.
